niri-taskbar: Initial Module Implementation

Nitepone opened the pull request:
This PR implements a taskbar that is specific for Niri. Unlike wlr-taskbar, this taskbar provides the ability to order applications by position in the Niri workspace; Groups applications by their workspace; and uses exclusively Niri IPC building on the implementation of the existing niri-workspaces module.

NOTE: This depends on an unmerged Niri PR, https://github.com/YaLTeR/niri/pull/1265

Nitepone commented:
Oh also, I have some aspirations to add support for multiple icon theme packs. And support for allowing configuration for characters to be used in place of icons.

These are enhancements that should probably crawl in through a later MR.

moving-hawk commented:
Hi, I'm an arch Linux user who wants to try niri/taskbar earlier.  I compiled the code using the PKGBUILD from the AUR after modifying the source URL.

It seems that `"WindowsLocationsChanged"` has been changed to `"WindowsLocationChanged"`, and `"tile_pos_in_scrolling_layout"` has been shortened to `"pos_in_scrolling_layout"`.

```[2025-09-02 13:02:15.107] [debug] Niri IPC: received {"WindowLayoutsChanged":{"changes":[[3,{"pos_in_scrolling_layout":[1,1],"tile_size":[2042.4,1079.2],"window_size":[2039,1076],"tile_pos_in_workspace_view":null,"window_offset_in_tile":[1.6,1.6]}],[2,{"pos_in_scrolling_layout":[2,1],"tile_size":[1020.0,1079.2],"window_size":[1017,1076],"tile_pos_in_workspace_view":null,"window_offset_in_tile":[1.6,1.6]}]]}}
```

When I connect an external display and close the laptop lid, Waybar stops running.

 ```
 [2025-09-02 13:14:57.183] [info] Bar configured (width: 1703, height: 30) for output: eDP-1

** (waybar:15151): ERROR **: 13:14:57.186:
unhandled exception (type unknown) in signal handler

Trace/breakpoint trap      (core dumped) waybar -c ~/.config/waybar/config_niri.jsonc -s ~/.config/waybar/style_niri.css -l debug
```

moving-hawk commented:
> > It seems that "WindowsLocationsChanged" has been changed to "WindowsLocationChanged", and "tile_pos_in_scrolling_layout" has been shortened to "pos_in_scrolling_layout".
> 
> This has been updated in my latest commit.

I believe I was using the latest commit (b0ad03163e7ae763dd6b67da9f9efb4ebc053d2a). If the two places mentioned below have been updated, I might have missed something.

```
diff --git a/src/modules/niri/taskbar.cpp b/src/modules/niri/taskbar.cpp
index 4f4dfd29..81ee0ca8 100644
--- a/src/modules/niri/taskbar.cpp
+++ b/src/modules/niri/taskbar.cpp
@@ -421,7 +421,7 @@ Taskbar::Taskbar(const std::string &id, const Bar &bar, const Json::Value &confi
 
   gIPC->registerForIPC("WindowsChanged", this);
   gIPC->registerForIPC("WindowOpenedOrChanged", this);
-  gIPC->registerForIPC("WindowsLocationsChanged", this);
+  gIPC->registerForIPC("WindowLayoutsChanged", this);
   gIPC->registerForIPC("WindowFocusChanged", this);
   gIPC->registerForIPC("WindowClosed", this);
 ```
 ```
diff --git a/src/modules/niri/backend.cpp b/src/modules/niri/backend.cpp
index 2a1bdb07..f353eef1 100644
--- a/src/modules/niri/backend.cpp
+++ b/src/modules/niri/backend.cpp
@@ -171,16 +171,16 @@ void IPC::parseIPC(const std::string &line) {
       windows_.clear();
       const auto &values = payload["windows"];
       std::copy(values.begin(), values.end(), std::back_inserter(windows_));
-    } else if (const auto &payload = ev["WindowsLocationsChanged"]) {
+    } else if (const auto &payload = ev["WindowLayoutsChanged"]) {
       // TODO THIS RELIES ON UNMERGED AN NIRI PR!!! CHECK AFTER IT IS MERGED https://github.com/YaLTeR/niri/pull/1265
       for (const auto &win_changes : payload["changes"]) {
         for (auto &win : windows_) {
           if (win["id"] == win_changes[0]) {
             // XXX We are deliberately dropping tile_size data for the sake of
             // ease of filtering spammed IPCs from moused based window resizing.
-            win["location"] = Json::Value(Json::objectValue);
-            win["location"]["tile_pos_in_scrolling_layout"]
-                = win_changes[1]["tile_pos_in_scrolling_layout"];
+            //win["location"] = Json::Value(Json::objectValue);
+            win["location"]["pos_in_scrolling_layout"]
+                = win_changes[1]["pos_in_scrolling_layout"];
             break;
           }
         }
```


> 
> > When I connect an external display and close the laptop lid, Waybar stops running
> 
> I have seen this a few times. I also recently saw it with the _other_ niri modules on mainline waybar. I suspect there is something up with niri backend setting some values to Null from a Niri IPC, then those are improperly handled by modules on bar refresh.
> 
> I swear I'll actually work on this over this week.. The joys of life becoming busy ^^'

It's fine. Life can be very busy owo. Thank you for your consideration!

Nitepone commented:
Oops, yeah it looks like I simply didn't properly test my original update to the IPCs. That should be fixed now, with a bonus of having fixed on of the intermittent crashes on monitor add/remove.

> Another question is how can I use the format 'icon-and-text'/'text'. My config doesn't work now.

This PR also needs docs for this. In lieu of that, here is a snippet from my config with some comments.

```
  "niri/taskbar": {
    // This sets the style of workspace groups.
    // Value is an enum of { "no-label", "label-idx", "label-ws-name" }
    "active-workspace-format" : "label-idx",
    "inactive-workspace-format" : "label-idx",

    // This sets the style of window buttons.
    // Value is an enum of { "icon", "text", "icon-and-text" }
    "active-button-format" : "icon-and-text",
    "inactive-button-format" : "icon",

    // The application icon size in px.
    "icon-size": 24,

    // The label used for the empty workspace indicator.
    "empty-ws-button-label": "+",
  },
```

As for css, the outer module element is named "taskbar" and each workspace has it's own names "workspace" div.

Thus css theming looks vaguely like.. 
```
#workspace {
  padding: 0 .4em;
  border-radius: .5em;
  background-color: #404040;
}

#workspace.active {
  background-color: #aa2020;
}

#workspace label {
  padding: 0 .4em;
}

#taskbar button {
  padding: 0 .2em;
  border-radius: .5em;
}

#taskbar button.active {
  background-color: #808080;
}

#taskbar button:hover {
  background: #2020aa;
}

#taskbar separator {
  background-color: #505050;
}
```

Nitepone commented:
Fixed that error message. Introduced it with my backend refactoring.

Updated the "button-format" enum to support using the window title instead of app id. "icon", "app-id", "title, "icon-and-app-id", "icon-and-title". ~~Though, the extremely long titles of some windows make this sometimes problematic.~~ I went ahead and added a new config property "label-max-length" to fix this issue. By default button labels are limited to 40 characters.

ADIX7 commented:
Is there a way to hide the new-workspace-button?

ADIX7 commented:
Also, the wlr/taskbar uses class `active` instead of `focused`, I don't know if it is intentional.
